### ISSUE TITLE
Golang: SessionOptions & ConfigProto opaqueness & inaccessibility

### ISSUE THREAD

**Issue by janhartman** [tensorflow/tensorflow#41699]
**System information**
- TensorFlow version (you are using): 1.15.0 (will move to stable 2.3.0 when it is available)
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
The Go bindings of TensorFlow have a somewhat awkward API when it comes to setting the session options: as per the [docs](https://pkg.go.dev/github.com/tensorflow/tensorflow@v1.15.0/tensorflow/go?tab=doc#SessionOptions), you can configure it, but you have to input a binary-serialized [ConfigProto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/config.proto) message to do so. 
However, you cannot actually create a ConfigProto by calling the Go TF bindings - you have to create it in another way, [as demonstrated in the source code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/go/session_test.go#L366). 
Creating the ConfigProto message by using Python is extremely impractical and unnecessarily opaque. I would urge the authors to reconsider using protocol buffers in the Go bindings to allow creating the ConfigProto message from Go, thus making this flow much simpler and user-transparent.

**Will this change the current API? How?**
This will introduce a dependency on protocol buffers in the Go bindings and add the generated protobuf structs to the API.

**Who will benefit with this feature?**
Users of the TensorFlow Go implementation with a workload that demands more configuration for effective inference.

**Comment by janhartman** [comment]
As a continuation: maybe even add support for other config options such as `RunOptions` - enable setting the run handler pool?

**Comment by jhollowayj** [comment]
So, I'm going to take a stab at what needs to happen for this.  Granted, I have no idea if this is right, but this is what I've gathered from poking around the codebase for an hour or so.

The config.proto file is defined [here](https://github.com/tensorflow/tensorflow/blob/c2081e9c26a713d86e5f6b68f9fa101d44e43d91/tensorflow/core/protobuf/config.proto).  If compiled via protoc and the go plugin, it should [generate code](https://github.com/tensorflow/tensorflow/blob/c2081e9c26a713d86e5f6b68f9fa101d44e43d91/tensorflow/core/protobuf/config.proto#L16) that will sit at github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf/for_core_protos_go_proto 
But I don't think that we are invoking the go/protoc plugin.  So now we jump into bazel.
I _think_ the bazel step to compile the config.proto file happens [here](https://github.com/tensorflow/tensorflow/blob/c2081e9c26a713d86e5f6b68f9fa101d44e43d91/tensorflow/core/protobuf/BUILD#L156-L186), since `config.proto` is included in the `COMMON_PROTO_SRCS` above in that file.
Now, that `tf_proto_library` function is defined way over [here](https://github.com/tensorflow/tensorflow/blob/48f0418e39203cf29c3508721f14062e4c30e0a3/tensorflow/core/platform/default/build_config.bzl#L493-L514].
There is now an parameter on line 509: `create_go_proto = False,`  Looking at the commit history for that commit, it seems like we never had a function implemented to call the protoc with go stuff at all.  

Here's where I start to get into assumptions. 

So, I think we need to add a new function `tf_proto_library_go` that mimics `tf_proto_library_cc`/`tf_proto_library_py`.  Each one of those seems to drop down into another function of `cc_proto_library`/`py_proto_library`.

The py_proto_library seems to reference a different repo with bazel stuff, somewhere in [this .bzl file of gh.com/protocolbuffers/protobuf](https://github.com/protocolbuffers/protobuf/blob/master/protobuf.bzl), so maybe there should be a go function in there?  I'm not sure.

Am I anywhere in the right ballpark, or am I way way off?
To me, it seems like it would be easier to bypass bazel build stuff and just compile the files as part of the Go setup stuff.  
_Something_ like the following? (it's getting late and I'm getting sleepy)

```
protoc \
  --proto_path=tensorflow/core/protobuf \
  --go_out=tensorflow/go/core/protobuf/for_core_protos_go_proto" \
  tensorflow/core/protobuf/*.proto
```

I didn't run any of this though, so it's all a stab in the dark.

**Comment by jhollowayj** [comment]
Ok, from the folder `github.com/tensorflow/tensorflow/tensorflow` (where the go folder is located), I can run

`protoc -I stream_executor/ -I core/framework/ -I core/protobuf/ --go_out=../../../ core/protobuf/config.proto`, but this gives errors.  I think that's also part of the problem.  But I'm giving up for now.  :)

**Comment by jhollowayj** [comment]
Last update.
I changed all of the `go_package` options in all of the protobuf files 
```diff
-option go_package = "github.com/tensorflow/tensorflow/tensorflow/go/core/framework/allocation_description_go_proto";
+option go_package = "tensorflow/go/core/framework";
```
Then I tried to do some protoc compiling, but I got some errors.  HOWEVER, I was able to generate the config go file via
```sh
cd /Users/jacob/go/src/github.com/tensorflow/tensorflow
mkdir -p ./tensorflow/go/core/protobuf/
protoc -I ./tensorflow/stream_executor/ -I ./tensorflow/core/framework/ -I ./tensorflow/core/protobuf/ -I ./ --go_out=./ tensorflow/core/protobuf/config.proto
```

This did have a `ConfigProto` structure in the generated file (i.e. `tensorflow/go/core/protobuf/config.pb.go`), so I _think_ that would then work.  
Perhaps we would want to make the `go_package` include another package for each protofile.  i.e. `tensorflow/go/core/framework/config`, but I don't have a clear answer on that one.
Also, are these generated files going to be committed, or do we have to generate them each time we want to build?  I don't know how that works with `go get` commands.

**Comment by ariden83** [comment]
Obviously, you should not use the official lib but that of wamuir: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/go

```
The TensorFlow team is not currently maintaning the Documentation for installing the Go bindings for TensorFlow.

The instructions has been maintained by the third party contributor: @wamuir

Please follow this [source](https://github.com/tensorflow/build/tree/master/golang_install_guide) by @wamuir for the installation of Golang with Tensorflow.
```

used : [wamuir > tensorflow ](https://github.com/wamuir/graft)

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
